First attempt:

Ekvo opened the pull request:
> modified: (go.mod,main.go,parcel.go,parcel_test.go,tracker.db);
> main.go ~> create connect to db sqlite, file - database;
> parcel.go ~> realise 6 functions:
> 1. func (s ParcelStore) Add(p Parcel) (int, error),
> 2. func (s ParcelStore) Get(number int) (Parcel, error),
> 3. func (s ParcelStore) GetByClient(client int) ([]Parcel, error),
> 4. func (s ParcelStore) SetStatus(number int, status string) error,
> 5. func (s ParcelStore) SetAddress(number int, address string) error,
> 6. func (s ParcelStore) Delete(number int) error;
> parcel_test.go ~> realise 4 test functions:
> 1. func TestAddGetDelete(t *testing.T),
> 2. func TestSetAddress(t *testing.T),
> 3. func TestSetStatus(t *testing.T),
> 4. func TestGetByClient(t *testing.T);
> tracker.db ~> rows is empty;

Commit:
First attempt:
> modified: (go.mod,main.go,parcel.go,parcel_test.go,tracker.db);
> main.go ~> create connect to db sqlite, file - database;
> parcel.go ~> realise 6 functions:
> 1. func (s ParcelStore) Add(p Parcel) (int, error),
> 2. func (s ParcelStore) Get(number int) (Parcel, error),
> 3. func (s ParcelStore) GetByClient(client int) ([]Parcel, error),
> 4. func (s ParcelStore) SetStatus(number int, status string) error,
> 5. func (s ParcelStore) SetAddress(number int, address string) error,
> 6. func (s ParcelStore) Delete(number int) error;
> parcel_test.go ~> realise 4 test functions:
> 1. func TestAddGetDelete(t *testing.T),
> 2. func TestSetAddress(t *testing.T),
> 3. func TestSetStatus(t *testing.T),
> 4. func TestGetByClient(t *testing.T);
> tracker.db ~> rows is empty;

ivannizh commented on parcel.go:
```diff
@@ -13,48 +14,139 @@ func NewParcelStore(db *sql.DB) ParcelStore {
+		}
+
+		res = append(res, p)
+	}
```
После цикла нужно проверить курсор на наличие ошибок

http://go-database-sql.org/retrieving.html

ivannizh commented on parcel_test.go:
```diff
@@ -30,58 +32,139 @@ func getTestParcel() Parcel {
-	db, err := // настройте подключение к БД
+	db, errDB := sql.Open("sqlite", "tracker.db")
+
+	require.Nil(t, errDB)
```
Есть специальная функция для ошибок
```suggestion
	require.NoError(t, err)
```

ivannizh replied:
Исправь, пожалуйста, везде

ivannizh commented on parcel_test.go:
```diff
@@ -30,58 +32,139 @@ func getTestParcel() Parcel {
+
 	// проверьте, что значения всех полей в полученном объекте совпадают со значениями полей в переменной parcel
+	// это невозможно т к number получаем во время записи в базу данных
+	wont := reflect.ValueOf(parcel)
```
Очень сложно, можно просто занулить значение, так как оно неважно
```suggestion
	stored.Number = 0
	require.Equal(t, parcel, stored)
```

ivannizh replied:
рефлексия тут точно не нужна

ivannizh commented on parcel_test.go:
```diff
@@ -30,58 +32,139 @@ func getTestParcel() Parcel {
+
 	// проверьте, что посылку больше нельзя получить из БД
+	_, errCheck := store.Get(id)
+	require.NotNil(t, errCheck)
```
Для сравнивания ошибок нужно использовать специализированную функцию
```suggestion
	require.ErrorIs(t, err, sql.ErrNoRows)
```

ivannizh commented on parcel_test.go:
```diff
@@ -108,14 +194,34 @@ func TestGetByClient(t *testing.T) {
 		// в parcelMap лежат добавленные посылки, ключ - идентификатор посылки, значение - сама посылка
 		// убедитесь, что все посылки из storedParcels есть в parcelMap
+		p, ex := parcelMap[parcel.Number]
+		assert.NotEmpty(t, ex)
```
```suggestion
		assert.True(t, ex)
```

ivannizh commented on parcel_test.go:
```diff
@@ -108,14 +194,34 @@ func TestGetByClient(t *testing.T) {
+		p, ex := parcelMap[parcel.Number]
+		assert.NotEmpty(t, ex)
+
+		wont := reflect.ValueOf(p)
```
Аналогично нужно заменить рефлексию на простое сравнение структур

ivannizh reviewed (commented):
Привет! Хорошая работа! Осталось исправить некоторые моменты и работа будет принята.

Commit:
Second attempt:
> modified: parcel.go, parcel_test.go tracker.db;
parcel.go ~> func (s ParcelStore) GetByClient(client int) ([]Parcel, error) now return res, rows.Err() at the end;
parcel_test.go ~> for check error now use require.NoError() or require.ErrorIs;

ivannizh reviewed (approved):
Отлично, принято!
